src/webgpu/shader/validation/expression/access/array.spec.ts

```diff
@@ -0,0 +1,243 @@
+export const description = `
+Validation tests for array access expressions
+
+* Index type
+* Result type
+* Early-evaluation errors
+`;
+
+import { makeTestGroup } from '../../../../../common/framework/test_group.js';
+import { keysOf, objectsToRecord } from '../../../../../common/util/data_tables.js';
+import { Type, kAllScalarsAndVectors, VectorType } from '../../../../util/conversion.js';
+import { ShaderValidationTest } from '../../shader_validation_test.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+g.test('index_type')
+  .desc('Tests valid index types for array access expressions')
+  .params(u =>
+    u.combine('type', [
+      'bool',
+      'u32',
+      'i32',
+      'abstract-int',
+      'f32',
+      'f16',
+      'abstract-float',
+      'vec2i',
+    ] as const)
+  )
+  .beforeAllSubcases(t => {
+    if (t.params.type === 'f16') {
+      t.selectDeviceOrSkipTestCase('shader-f16');
+    }
+  })
+  .fn(t => {
+    const ty = Type[t.params.type];
+    const enable = ty.requiresF16() ? 'enable f16;' : '';
+    const code = `${enable}
+    fn foo() {
+      var x = array(1,2,3);
+      let tmp = x[${ty.create(0).wgsl()}];
+    }`;
+    const expect =
+      t.params.type === 'i32' || t.params.type === 'u32' || t.params.type === 'abstract-int';
+    t.expectCompileResult(expect, code);
+  });
+
+const kTypes = objectsToRecord(kAllScalarsAndVectors);
+const kTypeKeys = keysOf(kTypes);
+
+g.test('result_type')
+  .desc('Tests that correct result type is produced for an access expression')
+  .params(u =>
+    u
+      .combine('type', kTypeKeys)
+      .combine('elements', [0, 4] as const)
+      .filter(t => {
+        const ty = kTypes[t.type];
+        if (ty instanceof VectorType) {
+          if (ty.elementType.kind === 'abstract-int' || ty.elementType.kind === 'abstract-float') {
+            return false;
+          }
+        } else {
+          if (ty.kind === 'abstract-int' || ty.kind === 'abstract-float') {
+            return false;
+          }
+        }
+        if (t.elements === 0) {
+          if (ty instanceof VectorType) {
+            if (ty.elementType.kind === 'bool') {
+              return false;
+            }
+          } else {
+            if (ty.kind === 'bool') {
+              return false;
+            }
+          }
```

Comment:
Could use `elementTypeOf(ty).kind === 'bool'` (or `elementTypeOf(ty) === Type.bool`), which handles the "vector or scalar" stuff for you.